core/src/main/java/org/elasticsearch/search/aggregations/pipeline/movavg/SimulatedAnealingMinimizer.java

```diff
@@ -0,0 +1,156 @@
+     *                      a cost per set of coefficeints
+     * @return              The new set of model coefficients
+     */
+    public static double[] minimize(MovAvgModel model, double[] startingCoef, EvictingQueue<Double> train, double[] test) {
+
+        double temp = 1;
+        double minTemp = 0.0001;
+        int iterations = 100;
+        double alpha = 0.9;
+
+        double[] bestCoef = startingCoef.clone();
+        double[] oldCoef = startingCoef.clone();
+        double[] newCoef;
+
+        double oldCost = cost(model, startingCoef, train, test);
+        double bestCost = oldCost;
+        double newCost;
+        double ap;
+
+        while (temp > minTemp) {
+            for (int i = 0; i < iterations; i++) {
+                newCoef = neighborCoef(oldCoef, temp);
+                newCost = cost(model, newCoef, train, test);
+
+                ap = acceptanceProbability(oldCost, newCost, temp);
+                if (ap > Math.random()) {
+                    oldCoef = newCoef.clone();
+                    oldCost = newCost;
+
+                    if (newCost < bestCost) {
+                        bestCost = newCost;
+                        bestCoef = newCoef.clone();
+                    }
+                }
+            }
+
+            temp *= alpha;
+
+            logger.debug("Temperature: " + temp + "  best cost: " + bestCost + " (" + StringUtils.join(bestCoef, ",") + ")");
+
+        }
+
+        logger.debug("Final Temperature: " + temp + "  best cost: " + bestCost + " (" + StringUtils.join(bestCoef, ",") + ")");
+        return bestCoef;
+    }
+
+    /**
+     * If the new cost is better than old, return 1.0.  Otherwise, return a double that increases
+     * as the two costs are closer to each other.
+     *
+     * @param oldCost   Old model cost
+     * @param newCost   New model cost
+     * @param temp      Current annealing temperature
+     * @return          The probability of accepting the new cost over the old
+     */
+    private static double acceptanceProbability(double oldCost, double newCost, double temp) {
+        return newCost < oldCost ? 1.0 : Math.exp(-(newCost - oldCost) / temp);
+    }
+
+    /**
+     * Generates a "neighboring" set of coefficients.  This is a set of coefficients
+     * where only one has been changed to try and isolate the effect of each parameter.
+     *
+     * @param coef  The old set of model coefficients
+     * @param temp  Current annealing temperature
+     * @return      A new set of coefficients where only one has changed
+     */
+    private static double[] neighborCoef(double[] coef, double temp) {
+        int toChange = (int)(Math.random() * (coef.length));
+        double[] newParams = coef.clone();
+        newParams[toChange] += (Math.random() * 2 - 1);         // TODO smarter choice dependent on temperature?
+        newParams[toChange] = Math.max(newParams[toChange], 0);
+        newParams[toChange] = Math.min(newParams[toChange], 1);
```

Comment:
My worry here is that if the coefficients hit (or get close to) 0 or 1 they have a 50% chance (or approaching 50% chance) of staying there, even if they are selected as the variable to change. In these cases, can we make sure the value moves from 1 or 0 rather than keeping it there?